Sources/PackageGraph/PinsStore.swift

```diff
@@ -84,93 +72,186 @@ public final class PinsStore {
     ///   - state: The state to pin at.
-    public func pin(
-        packageRef: PackageReference,
-        state: CheckoutState
-    ) {
-        self.pinsMap[packageRef.identity] = Pin(
+    public func pin(packageRef: PackageReference, state: CheckoutState) {
+        self.add(.init(
             packageRef: packageRef,
             state: state
-        )
+        ))
     }
 
     /// Add a pin.
     ///
     /// This will replace any previous pin with same package name.
     public func add(_ pin: Pin) {
-        self.pinsMap[pin.packageRef.identity] = pin
+        self._pins[pin.packageRef.identity] = pin
     }
 
     /// Unpin all of the currently pinned dependencies.
     ///
     /// This method does not automatically write to state file.
     public func unpinAll() {
         // Reset the pins map.
-        self.pinsMap = [:]
+        self._pins.clear()
     }
 
     public func saveState() throws {
-        if self.pinsMap.isEmpty {
+        try self.storage.save(pins: self._pins.get(), mirrors: self.mirrors, removeIfEmpty: true)
+    }
+}
+
+// MARK: - Serialization
+
+fileprivate struct PinsStorage {
+    private let path: AbsolutePath
+    private let fileSystem: FileSystem
+    private let encoder = JSONEncoder.makeWithDefaults()
+    private let decoder = JSONDecoder.makeWithDefaults()
+
+    init(path: AbsolutePath, fileSystem: FileSystem) {
+        self.path = path
+        self.fileSystem = fileSystem
+    }
+
+    func load(mirrors: DependencyMirrors) throws -> PinsStore.PinsMap {
+        if !self.fileSystem.exists(self.path) {
+            return [:]
+        }
+
+        // 👀 do we want a lock file here? safer but pretty noisy
+        //return try self.fileSystem.withLock(on: self.path, type: .shared) {
+            let version = try self.decoder.decode(path: self.path, fileSystem: self.fileSystem, as: Version.self)
+            switch version.version {
+            case 1:
+                let v1 = try decoder.decode(path: self.path, fileSystem: self.fileSystem, as: V1.self)
+                return try v1.object.pins.map{ try PinsStore.Pin($0, mirrors: mirrors) }.reduce(into: [PackageIdentity: PinsStore.Pin]()) { partial, iterator in
+                    if partial.keys.contains(iterator.packageRef.identity) {
+                        throw StringError("duplicated entry for package \"\(iterator.packageRef.name)\"")
+                    }
+                    partial[iterator.packageRef.identity] = iterator
+                }
+            default:
+                throw InternalError("unknown RepositoryManager version: \(version)")
+            }
+        //}
+    }
+
+    func save(pins: PinsStore.PinsMap, mirrors: DependencyMirrors, removeIfEmpty: Bool) throws {
+        if !self.fileSystem.exists(self.path.parentDirectory) {
+            try self.fileSystem.createDirectory(self.path.parentDirectory)
+        }
+        // 👀 do we want a lock file here? safer but pretty noisy
```

Comment:
Same as above.  Does it matter if it's noisy in the file system?  Is this the same as the `Package.resolved` file, or is this some place more hidden.  If it's `Package.resolved`, I see the point about noisiness...

Reply:
moved it into the workspace working directory (eg .build) so its more "quiet"